techmatters/terraso-mobile-client
# feat: offline soil

## Description

Rearchitect the Soil ID data loading system to support necessary offline behavior.
- Separate soil ID matches from the main soil ID slice
- Use a single app-wide context to coordinate loading soil ID matches in the background
- Selectively reload data-based matches when synced changes result in updated soil ID inputs, rather than flushing the entire cache any time a change action completes
- Flush the location-based match cache when the user switches from offline to online, to avoid retaining excess data

Also add a small selection of utility code to support this functionality, such as the `useValueSet` hook which tracks unique input values provided by child callbacks. Split the `soilId` model directory into `soilId`, `soilIdMatches`, and `soilMetadata` to avoid confusion caused by excessive size.

### Checklist

### Related Issues
#2372

### Verification steps

Verify that soil ID loads for temporary locations and sites in all existing contexts. Verify that data-based matches are reloaded when the user changes relevant inputs. Verify that data-based matches are reloaded when a *pull* changes relevant inputs. Verify that data-based matches matches are retained when the user goes offline, even between re-launching the app, and even if relevant inputs are changed by the user. Verify that data-based matches are reloaded for the most up-to-date inputs when the user goes back online.
